src/ProgressOnderwijsUtils/DistinctArray.cs

```diff
@@ -8,44 +8,98 @@ namespace ProgressOnderwijsUtils
+            => DistinctArray<T>.FromDistinctNonMutatedArray(items.ToArray(), EqualityComparer<T>.Default);
+
+        [Pure]
+        public static DistinctArray<T> ToDistinctArrayFromDistinct<T>([NotNull] this IEnumerable<T> items, IEqualityComparer<T> comparer)
+            => DistinctArray<T>.FromDistinctNonMutatedArray(items.ToArray(), comparer);
+
+        public static DistinctArray<T> ToDistinctArrayFromDistinct_Unchecked<T>([NotNull] this T[] items)
+            => DistinctArray<T>.FromDistinct_ClaimDistinctnessWithoutCheck(items);
     }
 
     [Serializable]
     public struct DistinctArray<T> : IReadOnlyList<T>
     {
         public static DistinctArray<T> Empty
-            => FromDistinct(Array.Empty<T>());
+            => new DistinctArray<T>(Array.Empty<T>());
 
-        public static DistinctArray<T> FromDistinct([NotNull] IEnumerable<T> items)
-            => FromDistinct(items, EqualityComparer<T>.Default);
+        public static DistinctArray<T> FromDistinct_ClaimDistinctnessWithoutCheck([NotNull] T[] items)
+            => new DistinctArray<T>(items);
 
-        public static DistinctArray<T> FromDistinct([NotNull] IEnumerable<T> items, IEqualityComparer<T> comparer)
+        public static DistinctArray<T> FromDistinctNonMutatedArray([NotNull] T[] items, IEqualityComparer<T> comparer)
         {
-            if (items.ContainsDuplicates(comparer)) {
+            var set = new HashSet<T>(items, comparer);
+            if (set.Count != items.Length) {
                 throw new ArgumentException("items are not distinct");
             }
-
-            return new DistinctArray<T> {
-                items = items.ToArray()
-            };
+            return new DistinctArray<T>(items);
         }
 
-        public static DistinctArray<T> FromPossiblyNotDistinct([NotNull] IEnumerable<T> items)
-            => FromPossiblyNotDistinct(items, EqualityComparer<T>.Default);
-
         public static DistinctArray<T> FromPossiblyNotDistinct([NotNull] IEnumerable<T> items, IEqualityComparer<T> comparer)
+            => new DistinctArray<T>(new HashSet<T>(items, comparer).ToArray());
+
+        readonly T[] items;
+        DistinctArray(T[] items) => this.items = items;
+
+        [NotNull]
+        public T[] UnderlyingArrayThatShouldNeverBeMutated() => items ?? Array.Empty<T>();
+
+        public int Count => UnderlyingArrayThatShouldNeverBeMutated().Length;
+        public T this[int index] => UnderlyingArrayThatShouldNeverBeMutated()[index];
+        public Enumerator GetEnumerator() => new Enumerator(UnderlyingArrayThatShouldNeverBeMutated());
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
+
+        /// <summary>
+        /// Efficient iterator for foreach.
+        /// </summary>
+        public struct Enumerator : IEnumerator<T>

```

Comment:
Moet deze public zijn?

Reply:
Ja, om boxing te voorkomen.

Reply:
ja, for foreach (dat doen ook alle builtin dingen als HashSet en Dictionary, dus het is ook niet ongewoon)